### PR TITLE
Allow installation of extra platform plugins

### LINKED ISSUE #88: Use of different platforms
Hi,
In [src/deployers/PlatformPluginsDeployer.cpp#L20](https://github.com/linuxdeploy/linuxdeploy-plugin-qt/blob/master/src/deployers/PlatformPluginsDeployer.cpp#L20) only `platforms/libqxcb.so` is copied over? Is there a way to get other platforms to ship? minimal, offscreen, vnc and alike?
Tried to add them via the `EXTRA_QT_PLUGINS` env variable but without success.

### PR DESCRIPTION
**Pull request by asl**
Resolves #88

### CODE REVIEW

**Comment by asl** [comment]
This is essentially code from https://github.com/linuxdeploy/linuxdeploy-plugin-qt/pull/89 but refactored to address review requests.

I checked that indeed it does deploy extra platform plugins and the final AppImage works

**Comment by asl** [comment]
Tagging @TheAssassin

**Comment by bjorn** [comment]
I'm just wondering, but why do we need `EXTRA_PLATFORM_PLUGINS` in addition to `EXTRA_QT_PLUGINS`? Are platform plugins not Qt plugins as well?

**Comment by asl** [comment]
@bjorn Apparently, yes. `EXTRA_QT_PLUGINS` is a pre-defined hardcoded list: https://github.com/linuxdeploy/linuxdeploy-plugin-qt/blob/dc2e5b7d648b6b46d9a80c8b78f8a38626b4976b/src/qt-modules.h#L21

Actually `EXTRA_QT_PLUGINS` name is a bit misleading, it should actually be `EXTRA_QT_MODULES`.

**Comment by asl** [comment]
@theassassin will you be able to review this? Thanks!

**Comment by asl** [comment]
@TheAssassin gently ping

**Comment by TheAssassin** [comment]
> Actually EXTRA_QT_PLUGINS name is a bit misleading, it should actually be EXTRA_QT_MODULES.

I think this was just taken over from linuxdeployqt. We should think about correcting this. Mind to open an issue?

File: ci/test.sh [hunk 1]
```diff
@@ -87,4 +87,15 @@ pushd linuxdeploy-plugin-qt-examples/QtWidgetsApplication
...
+
+    mkdir build-platforms
+    pushd build-platforms
+        export EXTRA_PLATFORM_PLUGINS="libqoffscreen.so;libqminimal.so"
+        qmake CONFIG+=release PREFIX=/usr ../QtWidgetsApplication.pro || exit 1
+        INSTALL_ROOT="$PWD"/AppDir make install || exit 1
+
+        "$LINUXDEPLOY_BIN" --appdir "$PWD"/AppDir --plugin qt --output appimage || exit 1
+        APPIMAGE=(*AppImage)
+        mv -v "${APPIMAGE[0]}" "$BUILD_DIR/offscreen-${APPIMAGE[0]}" || exit 1
```
Comment (TheAssassin) [inline ci/test.sh:99]: Maybe should check whether after the linuxdeploy call, the AppDir contains the requested files.

File: ci/test.sh [hunk 2]
```diff
@@ -87,4 +87,15 @@ pushd linuxdeploy-plugin-qt-examples/QtWidgetsApplication
...
         mv -v *AppImage "$build_dir" || exit 1
     popd
+
+    mkdir build-platforms
+    pushd build-platforms
+        export EXTRA_PLATFORM_PLUGINS="libqoffscreen.so;libqminimal.so"
+        qmake CONFIG+=release PREFIX=/usr ../QtWidgetsApplication.pro || exit 1
+        INSTALL_ROOT="$PWD"/AppDir make install || exit 1
+
+        "$LINUXDEPLOY_BIN" --appdir "$PWD"/AppDir --plugin qt --output appimage || exit 1
```
Comment (TheAssassin) [inline ci/test.sh:97]: It'll be easier to just `env OUTPUT=my.AppImage "$LINUXDEPLOY_BIN" [...]`. That way, you don't have to guess.

File: ci/test.sh [hunk 3]
```diff
@@ -87,4 +87,15 @@ pushd linuxdeploy-plugin-qt-examples/QtWidgetsApplication
...
+        APPIMAGE=(*AppImage)
```
Comment (TheAssassin) [inline ci/test.sh:98]: I guess it makes sense to use `*.AppImage` as a pattern. But generally, I'd prefer you to set the output filename (see above).

File: src/deployers/PlatformPluginsDeployer.cpp [hunk 4]
```diff
@@ -17,9 +18,19 @@ bool PlatformPluginsDeployer::deploy() {
 
     ldLog() << "Deploying platform plugins" << std::endl;
 
+    // always deploy default platform
     if (!appDir.deployLibrary(qtPluginsPath / "platforms/libqxcb.so", appDir.path() / "usr/plugins/platforms/"))
         return false;
 
+    // deploy extra platform plugins, if any
+    if (const auto* const platformPluginsFromEnvData = getenv("EXTRA_PLATFORM_PLUGINS")) {
```
Comment (TheAssassin) [inline src/deployers/PlatformPluginsDeployer.cpp:26]: I *hate* implicit comparisons, I really should introduce such a compiler flag. Could you please rewrite the code to check explicitly against `nullptr`?

**Review by TheAssassin — CHANGES_REQUESTED** [review]

File: src/deployers/PlatformPluginsDeployer.cpp (hunk 4 above)
Comment (asl) [reply src/deployers/PlatformPluginsDeployer.cpp]: Done

**Review by TheAssassin — APPROVED** [review]

**Comment by TheAssassin** [comment]
Haven't tested this extensively, but the code looks good and on a first try it seems to do the trick. Thanks @asl!

**Pull request merged**